articles/200_coexistence_and_migration.md

```diff
@@ -0,0 +1,247 @@
+
+#### Build type
+
+Almost all build types for ROS 1 and ROS 2 are disjoint except the build type `cmake` which can be processed by both build tools.
+Since it is not desired to change existing `catkin` / `ament_cmake` packages to use the build type `cmake` those are currently different for each ROS version.
+
+### CMake version
+
+ROS 2 requires CMake version 3.5+.
+It should be easily possible to update existing CMake code to be compatible with various versions of CMake.
+
+### Python version
+
+ROS 2 requires Python version 3.5+.
+It should be easily possible to update existing Python code to be compatible with Python 2.7 as well as Python 3.5+.
+
+### CMake API
+
+ROS 1 packages use the CMake API provided by [catkin](https://github.com/ros/catkin).
+ROS 2 packages use a slightly different CMake API provided by [ament_cmake](https://github.com/ament/ament_cmake).
+The rational for the changes and the different name are described in [this article](http://design.ros2.org/articles/ament.html).
+
+**Idea:** Create a ROS 2 package names `catkin` which provides the same CMake API as the ROS 1 package (minus unavailable features like the devel space) by mapping the calls to `ament_cmake` API.
+In addition it is expected that existing packages need to call a new function at the end of their CMake code to account for the fact that `ament_package` needs to be invoked *after* all targets where `catkin_package` needs to be invoked *before* them.
+
+### Message / service definitions
+
+In ROS 2 message / service definitions are expected to be in the subdirectory `msg` / `srv`.
+This implicitly ensures that message and service names are unique.
+There is no technical reason behind that design decision rather than simplifying the implementation.
+
+**Idea:** Allow arbitrary locations in the source space.
+
+Currently ROS 2 already supports additional feature in these files (e.g. default values) which are not available in ROS 1.
+In the future ROS 2 might support a different file format to satisfy the requested feature requests (like grouping constants in enums, optional fields, file / constant / field comments, etc.).
+If that is the case a backward compatible parser to read existing `.msg` / `.srv` files will likely be available.
+
+### Generated message / service API
+
+In order to satisfy the goal of dual-homed applications the generated code of ROS 1 and ROS 2 need to be side-by-side usable.
+
+For C++ that is already the case.
+In ROS 1 the header files are generated in namespace specific to the package name.
+In ROS 2 the header files are placed into a sub-namespace names `msg` / `srv`.
+Beside the different path / namespace the data structure also uses different types for e.g. the shared pointers (Boost vs. C++11).
+
+For Python that is not the case.
+
+**Proposal:** Change the location of the generated Python modules from using `msg` / `srv` as the subdirectory to `message` / `service`.
```

Comment:
I see no mention of a benefit for the rename.

Reply:
The benefit would be the ability to support dual-homed applications in Python (as mentioned a few lines above). Currently the generated Python files of ROS 1 and ROS 2 use the exact same path and can therefore not be used at the same time within a single application.

Reply:
I get it with your explanation. If it said: 'For python that is not the case, preventing packages from containing both ROS1 and ROS2 packages.' I would have understood the first time.

Reply:
Clarified in 937363a1969adc8a257d06c3c4eb6fc82e555a7a.